app/gui2/shared/ast/parse.ts

```diff
@@ -591,11 +596,121 @@ function resync(
+  // the logic for mapping raw types to `Ast` from the actual construction of the `Ast` types.
+  const parsed = abstract(module, rawParsed, code)
+  const newNodesBySpan = parsed.spans.nodes
+  const newSpans = new Map<AstId, SourceRange>()
+  for (const [key, asts] of parsed.spans.nodes) {
+    for (const ast of asts) newSpans.set(ast.id, sourceRangeFromKey(key))
+  }
+  return { parsedRoot: parsed.root, newSpans, newNodesBySpan, rawParsed }
+}
+
+function findIdenticalSubtrees(ast1: Ast, ast2: Ast) {
+  const ast1ByContent = hashTree(ast1)
+  const ast2ByContent = hashTree(ast2)
+  const ast1MatchedNodes = new Set<AstId>()
+  for (const [key, asts] of ast1ByContent.hashes) {
+    if (ast2ByContent.hashes.has(key)) {
+      for (const ast of asts) ast1MatchedNodes.add(ast.id)
+    }
+  }
+  const ast1MatchedSubtrees = new Set(subtreeRoots(ast1.module, ast1MatchedNodes))
+  const equivalences = new Array<readonly [AstId, AstId]>()
+  for (const [key, asts] of ast1ByContent.hashes) {
+    const ast1Roots = asts.filter((ast) => ast1MatchedSubtrees.has(ast.id))
+    const ast2Equivalents = ast2ByContent.hashes.get(key)
+    for (;;) {
+      const ast1Root = ast1Roots.shift()
+      if (!ast1Root) break
+      const ast2Root = ast2Equivalents!.shift()
+      if (!ast2Root) break
+      equivalences.push([ast1Root.id, ast2Root.id])
+    }
+  }
+  return equivalences
+}
+
+// TODO(#8238): `Module.applyTextEdits`:
+//  - Locate the change-roots by spans in the original tree.
+//  - `syncToCode` only the changed subtrees.
+export function syncToCode(ast: MutableAst, code: string) {
+  // Content-hash `ast` and `parsedCode` to find unchanged-roots.
+  const { parsedRoot, newSpans, rawParsed } = analyzeCode(
+    code,
+    ast instanceof BodyBlock,
+    ast.module,
+  )
+  const equivalences = findIdenticalSubtrees(ast, parsedRoot)
+  const partAfterToAstBefore = new Map<SourceRangeKey, Ast>()
+  for (const [astBefore, astAfter] of equivalences)
+    partAfterToAstBefore.set(
+      sourceRangeKey(newSpans.get(astAfter)!),
+      ast.module.checkedGet(astBefore),
+    )
+
+  const edit = ast.module
+  // `ast` itself may be in `subtreesToReuse`; if we `take` it when recycling subtrees, it would no longer have a place
+  // in the AST and we wouldn't be able to `replace` it later. Performing the whole edit in an `update` block ensures we
+  // can put the result in the right place.
+  ast.updateValue((_ast) => {
+    // Re-abstract from the root down to the unchanged-roots.
+    const recycledSubtrees = new Map(
+      Array.from(partAfterToAstBefore, ([span, ast]) => [
+        span as NodeKey,
+        edit.getVersion(ast).take(),
+      ]),
+    )
+    const parsedReusingSubtrees = abstract(edit, rawParsed, code, recycledSubtrees)
+    assertEqual(parsedReusingSubtrees.root.code(), code)
+    return parsedReusingSubtrees.root
+  })
```

Comment:
So, I guess from the algorithm, than changing code
`foo (2 + 2) bar () (2 + 2)` to `foo (2 + 2) bar (2 + 2) (2 + 2)` would shift the metadata from third argument to second argument? Is this something to be fixed by last PR?

Of course, it's rather a rare case, and we rather won't be able to properly handle all possible cases (GUI1 also does not).

Reply:
This will be handled in the next PR. The full text-sync algorithm uses an edit-aware primary heuristic, and only tries to pair-up identical but apparently-moved subtrees if there isn't a good edit-driven match.